mycodo/controllers/controller_trigger.py

```diff
@@ -317,36 +300,66 @@ def get_method_output(self, method_id):
         """ Get output variable from method """
         this_controller = db_retrieve_table_daemon(
             Trigger, unique_id=self.unique_id)
-        setpoint, ended = calculate_method_setpoint(
-            method_id,
-            Trigger,
-            this_controller,
-            Method,
-            MethodData,
-            self.logger)
+
+        now = datetime.datetime.now()
+
+        if this_controller.method_start_time is None or this_controller.method_start_time == 'Ready':
+            this_controller.method_start_time = now
+
+        method = load_method(method_id, self.logger)
+        setpoint, finished = method.calculate_setpoint(now, this_controller.method_start_time)
+
+        # Check if method_end_time is not None
+        end_time_reached = False
+        if this_controller.method_end_time:
+            # Convert time string to datetime object
+            end_time = datetime.datetime.strptime(
+                str(this_controller.method_end_time), '%Y-%m-%d %H:%M:%S.%f')
+            if now > end_time:
+                end_time_reached = True
+
+        if finished or end_time_reached:
+            # point in time is out of method range
+            with session_scope(MYCODO_DB_PATH) as db_session:
+                # Overwrite this_controller with committable connection
+                this_controller = db_session.query(Trigger).filter(
+                    Trigger.unique_id == self.unique_id).first()
+
+                if not end_time_reached and method.should_restart(now, this_controller.method_start_time):
+                    self.logger.debug("Restart")
+                    # Method has been instructed to restart
+                    this_controller.method_start_time = now
+                    # Calculate restart value
+                    setpoint, finished = method.calculate_setpoint(now, this_controller.method_start_time)
+                else:
+                    self.logger.debug("Ended")
+                    # Duration method has ended, reset method_start_time locally and in DB
+                    this_controller.method_start_time = 'Ended'
+                    this_controller.method_end_time = None
+                    this_controller.is_activated = False
+                    db_session.commit()
+
+                    self.is_activated = False
+                    self.stop_controller()
+
+                db_session.commit()
```

Comment:
Previous implementation did database access to controller tables partly in the controller and partly in the method. The part that was done in method has been migrated to the controllers.